util/include/MultiSizeBufferPool.hpp

```diff
@@ -0,0 +1,483 @@
+
+class MultiSizeBufferPoolMetrics {
+ public:
+  MultiSizeBufferPoolMetrics(const std::string& name)
+      : component_{std::string("MultiSizeBufferPoolMetrics") + name, std::make_shared<concordMetrics::Aggregator>()},
+        overallUsedBuffersFailure_{component_.RegisterAtomicGauge("overallUsedBuffersFailure", 0)},
+        overallUsedBuffersSuccess_{component_.RegisterAtomicGauge("overallUsedBuffersSuccess", 0)},
+        overallUsedBuffers_{component_.RegisterAtomicGauge("overallUsedBuffers", 0)},
+        overallMaxUsedBytes_{component_.RegisterAtomicGauge("overallMaxUsedBytes", 0)},
+        currentUsedBytes_{component_.RegisterAtomicGauge("currentUsedBytes", 0)} {
+    component_.Register();
+  }
+
+  void setAggregator(const std::shared_ptr<concordMetrics::Aggregator>& aggregator) {
+    component_.SetAggregator(aggregator);
+  }
+  void updateAggregator() { component_.UpdateAggregator(); }
+  concordMetrics::Component component_;
+
+  // metrics handles
+
+  // Number of times a buffer was given to caller, but caller couldn't used it since buffer wasn't big enough
+  concordMetrics::AtomicGaugeHandle overallUsedBuffersFailure_;
+  // Number of times a buffer was given to caller and could be used
+  concordMetrics::AtomicGaugeHandle overallUsedBuffersSuccess_;
+  // overallUsedBuffersFailure_+ overallUsedBuffersSuccess_
+  concordMetrics::AtomicGaugeHandle overallUsedBuffers_;
+  // Maximum sum of bytes used by all callers: sum of all used buffers
+  concordMetrics::AtomicGaugeHandle overallMaxUsedBytes_;
+  // current sum of bytes used by all callers: sum of all used buffers
+  concordMetrics::AtomicGaugeHandle currentUsedBytes_;
+};
+
+namespace test {
+class MultiSizeBufferPoolTestFixture;
+}
+
+class MultiSizeBufferPool {
+  friend class concordUtil::test::MultiSizeBufferPoolTestFixture;  // Should be strictly used for testing!
+  friend class IPurger;
+
+ public:
+  // TODO - write full explanation for each member
+  struct SubpoolConfig {
+    const uint32_t bufferSize;         // size of data which can be used by a client
+    const uint32_t numInitialBuffers;  // number of buffers to allocate on creation
+    const uint32_t numMaxBuffers;      // maximum number of buffers that can be allocated (initially + dynamically)
+  };
+  using SubpoolsConfig = std::vector<MultiSizeBufferPool::SubpoolConfig>;
+
+  struct Config {
+    // The default timer interval, in seconds
+    static constexpr uint32_t kDefaultBasePeriodicTimerIntervalSec = 1;
+
+    // When maxAllocatedBytes is zero - do not enforce a limit on pool total size
+    // when non-zero - sum of initial sizes should never exceed this parameter
+    // This parameter takes into account the net buffer size, without chunk header
+    // includes the metadata. Per each buffer, chunkMetadataSize() bytes are used (small amount)
+    const uint64_t maxAllocatedBytes;
+    // When true, throws an exception when a leak is suspected
+    const bool WarnOnSuspectedLeak;
```

Comment:
is there a test that intentionally causes buffer leaks  in order to check the leak detection and exception functionality?

Reply:
comment: this parameter has been removed, i always warn in log. no way to throw.